org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/BaseDocumentLifeCycleHandler.java

```diff
@@ -549,10 +563,75 @@ private ICompilationUnit checkPackageDeclaration(String uri, ICompilationUnit un
 			} catch (CoreException e) {
 				JavaLanguageServerPlugin.logException(e.getMessage(), e);
 			}
+		} else if (!ProjectUtils.isVisibleProject(project)) {
+			PreferenceManager preferencesManager = JavaLanguageServerPlugin.getPreferencesManager();
+			List<String> sourcePaths = preferencesManager.getPreferences().getInvisibleProjectSourcePaths();
+
+			// user already set the source paths manually, we don't infer it anymore.
+			if (sourcePaths != null) {
+				return unit;
+			}
+
+			boolean needToInfer = needInferSourceRoot(javaProject, unit);
+			if (!needToInfer) {
+				return unit;
+			}
+
+			IPath unitPath = unit.getResource().getLocation();
+			InvisibleProjectImporter.inferSourceRoot(javaProject, unitPath);
 		}
 		return unit;
 	}
 
+	/**
+	 * Checks if it's necessary to infer a source root based on the compilation unit.
+	 * Returns true when:
+	 * <ul>
+     *   <li>The compilation unit is not on the classpath, but belongs to the Java project.</li>
+     *   <li>The compilation unit is on the classpath, and all the compilation units of its belonging
+	 *       package fragment have {@value IProblem#PackageIsNotExpectedPackage} errors.</li>
+     * </ul>
+	 * @param javaProject Java project.
+	 * @param unit compilation unit.
+	 */
+	private boolean needInferSourceRoot(IJavaProject javaProject, ICompilationUnit unit) {
+		if (javaProject.isOnClasspath(unit)) {
+			IJavaElement parent = unit.getParent();
+			if (parent == null || !(parent instanceof IPackageFragment)) {
+				return false;
+			}
```

Comment:
We may quickly check whether the focus unit contains PackageIsNotExpectedPackage problem first. If it's true, then further check the peers.